tests/test_base.py

```diff
@@ -113,31 +113,31 @@ def test_format_results_for_report():
                 'line_number': 4,
                 'annotation_token': 'not_in_a_group',
                 'annotation_data': 'file 1 annotation 3',
-                'expected_group_id': None
+                'expected_group_id': 2
```

Comment:
The refactoring is not 100% backward-compatible: the previous implementation filtered out annotations that did not belong to a real group. I believe (though I'm not completely sure) this is because the parser used to capture unrelated annotations. I *think* this should no longer be the case since we improved the parser regular expression, in a previous PR.

However, in practice, and according to my tests, the new implementation is completely backward-compatible on edx-platform.

Reply:
I’m not clear how unrelated an unrelated toggle would be. I’ve been wondering how and if we detect bad annotations. My concern is that now that we have optional annotations, would we provide no complaints whether someone uses `.. toggle_warning` or `.. toggle_warnings`?

Happy to help brainstorm solutions after confirming and agreeing:
1. How it works.
2. What we want.

Reply:
I understand your confusion, it's a tricky piece of code. You should remember that here, we are "only" modifying the report generation code -- not the actual linting of annotations.
To be honest, I'm not even sure where this report generation code is actually being used. But it's easy enough to do some testing to check whether it matches your expectations:

```
$ cat foo.py
"""
.. toggle_name: MyToggle
.. toggle_warning: Warning 1
.. toggle_warnings: Warning 2
.. toggle_implementation: WaffleFlag
.. toggle_default: 42
.. toggle_description: Some description
.. toggle_use_cases: temporary
.. toggle_creation_date: 2000 AD
"""
$ rm -rf reports \
    && code_annotations static_find_annotations --config_file ./code_annotations/contrib/config/feature_toggle_annotations.yaml --source_path ./foo.py -vv
...
Performing linting checks...                                                                                                                                                                                       
{  'foo.py': [  {  'annotation_data': 'MyToggle',                                                                                                                                                                  
                   'annotation_token': '.. toggle_name:',                                                                                                                                                          
                   'filename': 'foo.py',                                                                                                                                                                           
                   'found_by': 'python',                                                                                                                                                                           
                   'line_number': 1},                                                                                                                                                                              
                {  'annotation_data': 'Warning 2',                                                                                                                                                                 
                   'annotation_token': '.. toggle_warnings:',                                                                                                                                                      
                   'filename': 'foo.py',                                                                                                                                                                           
                   'found_by': 'python',                                                                                                                                                                           
                   'line_number': 1},                                                                                                                                                                              
                {  'annotation_data': ['WaffleFlag'],                                                                                                                                                              
                   'annotation_token': '.. toggle_implementation:',                                                                                                                                                
                   'filename': 'foo.py',                                                                                                                                                                           
                   'found_by': 'python',                                                                                                                                                                           
                   'line_number': 1},                                                                                                                                                                              
                {  'annotation_data': '42',                                                                                                                                                                        
                   'annotation_token': '.. toggle_default:',                                                                                                                                                       
                   'filename': 'foo.py',                                                                                                                                                                           
                   'found_by': 'python',                                                                                                                                                                           
                   'line_number': 1},                                                                                                                                                                              
                {  'annotation_data': 'Some description',                                                                                                                                                          
                   'annotation_token': '.. toggle_description:',                                                                                                                                                   
                   'filename': 'foo.py',                                                                                                                                                                           
                   'found_by': 'python',                                                                                                                                                                           
                   'line_number': 1},                                                                                                                                                                              
                {  'annotation_data': ['temporary'],                                                                                                                                                               
                   'annotation_token': '.. toggle_use_cases:',                                                                                                                                                     
                   'filename': 'foo.py',                                                                                                                                                                           
                   'found_by': 'python',                                                                                                                                                                           
                   'line_number': 1},                                                                                                                                                                              
                {  'annotation_data': '2000 AD',                                                                                                                                                                   
                   'annotation_token': '.. toggle_creation_date:',                                                                                                                                                 
                   'filename': 'foo.py',                                                                                                                                                                           
                   'found_by': 'python',                                                                                                                                                                           
                   'line_number': 1}]}

Linting passed without errors.
Writing report...
foo.py:
- annotation_data: MyToggle
  annotation_token: '.. toggle_name:'
  filename: foo.py
  found_by: python
  line_number: 1
- annotation_data: Warning 2
  annotation_token: '.. toggle_warnings:'
  filename: foo.py
  found_by: python
  line_number: 1
 annotation_data:
  - WaffleFlag
  annotation_token: '.. toggle_implementation:'
  filename: foo.py
  found_by: python
  line_number: 1
- annotation_data: '42'
  annotation_token: '.. toggle_default:'
  filename: foo.py
  found_by: python
  line_number: 1
- annotation_data: Some description
  annotation_token: '.. toggle_description:'
  filename: foo.py
  found_by: python
  line_number: 1
- annotation_data:
  - temporary
  annotation_token: '.. toggle_use_cases:'
  filename: foo.py
  found_by: python
  line_number: 1
- annotation_data: 2000 AD
  annotation_token: '.. toggle_creation_date:'
  filename: foo.py
  found_by: python
  line_number: 1

report_format: formatting foo.py
Starting group id 1 for "[{'found_by': 'python', 'filename': 'foo.py', 'line_number': 1, 'annotation_token': '.. toggle_name:', 'annotation_data': 'MyToggle'}, {'found_by': 'python', 'filename': 'foo.py', 'line_
number': 1, 'annotation_token': '.. toggle_warnings:', 'annotation_data': 'Warning 2'}, {'found_by': 'python', 'filename': 'foo.py', 'line_number': 1, 'annotation_token': '.. toggle_implementation:', 'annotation
_data': ['WaffleFlag']}, {'found_by': 'python', 'filename': 'foo.py', 'line_number': 1, 'annotation_token': '.. toggle_default:', 'annotation_data': '42'}, {'found_by': 'python', 'filename': 'foo.py', 'line_numb
er': 1, 'annotation_token': '.. toggle_description:', 'annotation_data': 'Some description'}, {'found_by': 'python', 'filename': 'foo.py', 'line_number': 1, 'annotation_token': '.. toggle_use_cases:', 'annotatio
n_data': ['temporary']}, {'found_by': 'python', 'filename': 'foo.py', 'line_number': 1, 'annotation_token': '.. toggle_creation_date:', 'annotation_data': '2000 AD'}]" token ".. toggle_name:", line 1
report_format: Adding .. toggle_name: to group id 1
report_format: Adding .. toggle_warnings: to group id 1
report_format: Adding .. toggle_implementation: to group id 1
report_format: Adding .. toggle_default: to group id 1
report_format: Adding .. toggle_description: to group id 1
report_format: Adding .. toggle_use_cases: to group id 1
report_format: Adding .. toggle_creation_date: to group id 1
Generating report to reports/2021-30-03-15-15-43.yaml
Report written to reports/2021-30-03-15-15-43.yaml.
Search found 7 annotations in 0:00:00.023914.
$ cat reports/*.yaml
foo.py:
- annotation_data: MyToggle
  annotation_token: '.. toggle_name:'
  filename: foo.py
  found_by: python
  line_number: 1
  report_group_id: 1
- annotation_data: Warning 2
  annotation_token: '.. toggle_warnings:'
  filename: foo.py
  found_by: python
  line_number: 1
  report_group_id: 1
- annotation_data:
  - WaffleFlag
  annotation_token: '.. toggle_implementation:'
  filename: foo.py
  found_by: python
  line_number: 1
  report_group_id: 1
- annotation_data: '42'
  annotation_token: '.. toggle_default:'
  filename: foo.py
  found_by: python
  line_number: 1
  report_group_id: 1
- annotation_data: Some description
  annotation_token: '.. toggle_description:'
  filename: foo.py
  found_by: python
  line_number: 1
  report_group_id: 1
- annotation_data:
  - temporary
  annotation_token: '.. toggle_use_cases:'
  filename: foo.py
  found_by: python
  line_number: 1
  report_group_id: 1
- annotation_data: 2000 AD
  annotation_token: '.. toggle_creation_date:'
  filename: foo.py
  found_by: python
  line_number: 1
  report_group_id: 1
```

We can also check edx-lint reports:
```
$ pip install edx-lint
$ pylint ./foo.py 
************* Module foo
foo.py:1:0: C0102: Black listed name "foo" (blacklisted-name)
foo.py:1:0: E7662: temporary feature toggle (MyToggle) has no target removal date (toggle-missing-target-removal-date)
foo.py:1:0: E7663: feature toggle (MyToggle) default value must be boolean ('True' or 'False') (toggle-non-boolean-default-value)
```

Happy to discuss whether and how we should improve this report. We should just keep in mind that we are relying on crude regular expressions to parse the source code. If we want "smarter" checks, we will probably have to switch to grammar parsing tools.

Reply:
Thanks.  As discussed, let's merge this and I'll document a future iteration that could detect unwanted annotations by potentially removing the ability to add comments in-between annotations as seen in this weird test case: https://github.com/edx/code-annotations/blob/master/tests/extensions/python_test_files/multiline_singlelinecomment.pyt